test/dynamo/test_repros.py

```diff
@@ -2987,10 +2987,37 @@ def f(x):
 
         self.assertEqual(f(torch.ones(8, 4)), gm(torch.ones(8, 4)))
 
+    def test_optim_state_references_cleared(self):
+        model = torch.nn.Linear(2048, 2048, bias=False)
+        x = torch.ones(2048)
+        state_ref = 0
+
+        optimizer = torch.optim.Adadelta(model.parameters(), lr=0.01)
+
+        def opt_step():
+            optimizer.step()
+
+        compiled_opt_step = torch._dynamo.optimize("eager")(opt_step)
+
+        def compiled_model_step(x):
+            optimizer.zero_grad(True)
+            y = model(x)
+            torch.sum(y).backward()
+            compiled_opt_step()
+
+        compiled_model_step(x)
+
+        state_ref = weakref.ref(
+            optimizer.state[optimizer.param_groups[0]["params"][0]]["square_avg"]
```

Comment:
Is this an arbitrary choice to just show that we're not holding onto any extra tensors the optimizer associates with each param? If so, this could be fragile where if square_avg changes its name or something, this test would just silently pass.

Reply:
Yes it is. Wouldn't we get a key not found error in that case? I do agree this is fragile. I'm not sure I have a better idea though.

Reply:
Would adding this line before the `optimizer = None` work?
```
self.assertTrue(state_ref() is not None)
```

Reply:
Discussed offline - it isn't really possible to create a weakref to an already dead object, I will add a comment clarifying that the param is an arbitrary choice